woocommerce/woocommerce
# Add PhoneNumberInput component – cooldown

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.


### Changes proposed in this Pull Request:
After wrangling a phone number input components for WooPayments, it was suggested to move it to WooCommerce (paJDYF-85M-p2#comment-17423). I also asked for a confirmation about the implementation here (p1687780655339429/1687778544.623139-slack-C01SFMVEYAK).

The reason to add this new component is to remove [`intl-tel-input`](https://github.com/jackocnr/intl-tel-input) from WooPayments, to hugely reduce the bundle size and have a React based customizable component.

#### Main features
- Metadata (minimum required information)
  - Built by manually running `bin/packages/js/components/phone-number-input/build-data.js` with data coming from `intl-tel-input` and `phone` packages.
  - Country names extracted from `window.wcSettings`.
  - Flags come from `s.w.org` twemoji images, using the country code to generate the required image name for the URL.
- Component
  - Custom select control made using `downshift` which allows selecting a country by mouse/keyboard picking it from the menu or typing the name while it’s focused.
  - Bare minimum style, wrapping everything with an input and with the optional prop `className` to make it easier to customize.
  - Can be attached to a label with the optional prop `id` .
  - Includes 3 render function props to easily override how the selected value, menu items, and menu arrow looks.
  - The country is guessed from the initial value, persisting the picked one after that.
  - Allows input of numbers, spaces, and hyphens to make it easier to write readable phone numbers.
  - On change, it returns the unaltered value, an E164 version of it, and the country alpha2 code.
- Validation
  - Sanitizes phone number to leave only the plus sign and numbers.
  - Returns early if the format doesn’t match E164.
  - Guess the country from the phone number if it’s not provided.
  - Checks the length if the country has it defined.
  - Checks the start if the country has it defined.

### How to test the changes in this Pull Request:
1. Build storybook – `cd tools/storybook && npm run storybook`.
2. Go to `http://localhost:6007/`
3. Pick `PhoneNumberInput` from the sidebar.
4. Interact with the component to check that it works as expected.
5. You can also play with the story code in `packages/js/components/src/phone-number-input/stories/index.tsx`.


### Screenshots
<img width="225" alt="Screenshot 2023-09-21 at 14 48 41" src="https://github.com/woocommerce/woocommerce/assets/7670276/ec088caa-e839-47d9-8c99-a58ba070a07c">
<img width="223" alt="Screenshot 2023-09-21 at 14 49 07" src="https://github.com/woocommerce/woocommerce/assets/7670276/90f543fb-5541-4dc4-8b10-cb494b51d25d">


### Changelog entry




<details>

#### Significance




#### Type




#### Message 
An international phone number input with country selection, and mobile phone numbers validation.

#### Comment 

</details>
